0000-settlement-engine/0000-settlement-engine.md

```diff
@@ -0,0 +1,353 @@
+
+1. Alice sends a custom instruction to her settlement engine to settle the fee. Note that she does not use the instruction reserved for the accounting system.
+2. Alice's settlement engine automatically negotiates the fee with Bob's settlement engine (likely subject to some limit).
+3. Alice's settlement engine sends a payment on the shared ledger or system to Bob's settlement engine for the value of the fee.
+4. Bob's settlement engine, instead of informing the accounting system of the incoming settlement, does nothing, but internally tracks that the negotiated fee was paid.
+5. If Alice's accounting system triggers any subsequent settlements to Bob, Bob's settlement engine should acknowledge them to the accounting system as normal.
+
+## Specification
+
+Operating a settlement engine requires three components each exposing a particular HTTP API: a settlement engine, an accounting system, and a transport.
+
+API endpoints exposed by these systems may be classified in two ways:
+
+1. Endpoints that may be invoked by one of these other two components, and MUST NOT be invoked manually. At a minimum, all settlement engine implementations MUST support these endpoints.
+2. Endpoints that may be invoked manually by an operator.
+
+This document outlines the first class of endpoints: those only essential to automated interoperability between these systems, and not manual operation.
+
+### Accounts and identifiers
+
+Each account MUST be identified by a unique, URL-safe string.
+
+#### Accounts vs ledger identities
+
+The settlement engine is RECOMMENDED to be responsible for correlating an account identifier to the peer's identity on the shared ledger or settlement system, if required.
+
+For separation of concerns between the accounting system and the settlement system, the accounting system is NOT RECOMMENDED to have knowledge of the peer's identity on the shared settlement system, but rather only the account identifier described here.
+
+#### Multiple accounts
+
+Settlement engine implementations SHOULD support settling with multiple accounts.
+
+#### Correlating messages
+
+The transport MUST correlate incoming messages from the peer to the correct account identifier, and MUST correlate outgoing messages from an account identifier to the correct peer to send them to.
+
+### Units and quantities
+
+#### `Quantity` type
+
+A quantity represents an amount denominated in some unit of a particular fungible asset. In uses of this type, the asset should be implicit.
+
+##### Attributes
+
+- **`amount`** string
+  - Number of the given unit of the asset between 0 and the maximum 64-bit unsigned integer, 18,446,744,073,709,551,615 (inclusive).
+  - The amount is encoded as a string to ensure no precision is lost on platforms that don't natively support full precision 64-bit integers.
+- **`scale`** number
+  - Number of orders of magnitude smaller this unit is as compared to the unit of exchange for the asset.
```

Comment:
> Number of orders of magnitude smaller this unit is as compared to the unit of exchange for the asset.

This is a bit unclear to me. What about this definition (which is adapted from [LPIv1](https://github.com/interledger/rfcs/blob/master/deprecated/0004-ledger-plugin-interface/0004-ledger-plugin-interface.md))

**`scale`** number
The order of magnitude used to express one full asset unit in the base units of the underlying settlement account. More formally, an integer (..., -2, -1, 0, 1, 2, ...), such that one full unit of a particular asset equals `10^(-scale)` units in the underlying account.

For example, if an account is denominated in "usd-cents", then the account's scale would be 2. If a caller supplies an `amount` of 100 units (i.e., 100 usd-cents), then this amount could be translated into 1 base units (i.e., dollars) via the following equation: `100 * (10^(-2)) = 1.00`.
```

Reply:
Agree. One question I have is, if the asset is U.S. dollars, what is the _unit_ for $1 called? I use "unit of exchange;" those examples use "full asset unit" and "base units." I don't think any of those terms are very clear. (And it's really hard to find any good references on the nomenclature a domain expert uses!) In any case, I think we need to more precisely define that concept and others.

Reply:
This has me thinking we should more clearly define that there are two layers (ledgers, if you will), both dealing with unit values. At the ILP layer, there is a "**Clearing Ledger**" that has a unit of account. In the Settlement Engine layer, there' is a "**Settlement Ledger**" that also has a unit of account. 

As an example, imagine we have an ILP account that denominates in cents (USD). Occasionally, we settle once one of us owes the other $1.00 (100 cents). In this example, the "Settlement Ledger" units would be dollars, and the "Clearing Ledger" units would be "cents."

Given those two definitions (Clearing and Settlement ledgers), I think I would tweak the definition of `scale` to be as follows:

_Scale: The power of 10 that one unit in the Clearing Ledger should be multiplied by in order to express the same value in the units of a corresponding Settlement Ledger. More formally, an integer (..., -2, -1, 0, 1, 2, ...), such that one unit in the Clearing Ledger equals 10^(-scale) units in the Settlement Ledger._

Reply:
Resolving this since I left a different comment that builds upon your good updates, @kincaidoneil